diff -Nur numpy-1.9.0.orig/numpy/core/include/numpy/npy_cpu.h numpy-1.9.0/numpy/core/include/numpy/npy_cpu.h
--- numpy-1.9.0.orig/numpy/core/include/numpy/npy_cpu.h	2015-05-18 13:29:53.800166857 -0700
+++ numpy-1.9.0/numpy/core/include/numpy/npy_cpu.h	2015-05-18 13:30:33.288237334 -0700
@@ -75,6 +75,8 @@
     #define NPY_CPU_AARCH64
 #elif defined(__mc68000__)
     #define NPY_CPU_M68K
+#elif defined(__riscv__)
+    #define NPY_CPU_RISCV
 #else
     #error Unknown CPU, please report this to numpy maintainers with \
     information about your platform (OS, CPU and compiler)
