diff --git a/sysdeps/riscv/preconfigure b/sysdeps/riscv/preconfigure
index e118a5ed28f6..cfe69ca187da 100644
--- a/sysdeps/riscv/preconfigure
+++ b/sysdeps/riscv/preconfigure
@@ -1,49 +1,8 @@
 case "$machine" in
 riscv*)
-    xlen=`$CC $CFLAGS $CPPFLAGS -E -dM -xc /dev/null | sed -n 's/^#define __riscv_xlen \(.*\)/\1/p'`
-    flen=`$CC $CFLAGS $CPPFLAGS -E -dM -xc /dev/null | sed -n 's/^#define __riscv_flen \(.*\)/\1/p'`
-    float_abi=`$CC $CFLAGS $CPPFLAGS -E -dM -xc /dev/null | sed -n 's/^#define __riscv_float_abi_\([^ ]*\) .*/\1/p'`
-    
-    case "$xlen" in
-    32 | 64)
-    	;;
-    *)
-    	echo "Unable to determine XLEN" >&2
-    	exit 1
-    	;;
-    esac
-    
-    case "$flen" in
-    64)
-    	float_machine=rvd
-    	;;
-    32)
-    	float_machine=rvf
-    	;;
-    "")
-    	float_machine=
-    	;;
-    *)
-    	echo "Unable to determine FLEN" >&2
-    	exit 1
-    	;;
-    esac
-    
-    case "$float_abi" in
-    soft)
-    	abi_flen=0
-    	;;
-    single)
-    	abi_flen=32
-    	;;
-    double)
-    	abi_flen=64
-    	;;
-    *)
-    	echo "Unable to determine floating-point ABI" >&2
-    	exit 1
-    	;;
-    esac
+    xlen=64
+    flen=64
+    float_abi=double
     
     base_machine=riscv
     machine=riscv/rv$xlen/$float_machine
diff --git a/sysdeps/unix/sysv/linux/riscv/configure b/sysdeps/unix/sysv/linux/riscv/configure
index 4dd085a7b0b6..951e770ac623 100755
--- a/sysdeps/unix/sysv/linux/riscv/configure
+++ b/sysdeps/unix/sysv/linux/riscv/configure
@@ -158,9 +158,9 @@ if (eval "$ac_cpp conftest.$ac_ext") 2>&5 |
 fi
 rm -f conftest*
 
-if test $libc_cv_riscv_int_abi = no; then
-  as_fn_error $? "Unable to determine integer ABI" "$LINENO" 5
-fi
+#if test $libc_cv_riscv_int_abi = no; then
+#  as_fn_error $? "Unable to determine integer ABI" "$LINENO" 5
+#fi
 
 libc_cv_riscv_float_abi=no
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
@@ -203,7 +203,8 @@ fi
 rm -f conftest*
 
 if test $libc_cv_riscv_float_abi = no; then
-  as_fn_error $? "Unable to determine floating-point ABI" "$LINENO" 5
+#  as_fn_error $? "Unable to determine floating-point ABI" "$LINENO" 5
+  libc_cv_riscv_float_abi=d
 fi
 
 config_vars="$config_vars
diff --git a/sysdeps/unix/sysv/linux/riscv/shlib-versions b/sysdeps/unix/sysv/linux/riscv/shlib-versions
index fe7322d23a94..d8b968209933 100644
--- a/sysdeps/unix/sysv/linux/riscv/shlib-versions
+++ b/sysdeps/unix/sysv/linux/riscv/shlib-versions
@@ -13,5 +13,5 @@ ld=ld-linux-riscv32-ilp32f.so.1
 %elif RISCV_ABI_XLEN == 32 && RISCV_ABI_FLEN == 0
 ld=ld-linux-riscv32-ilp32.so.1
 %else
-%error cannot determine ABI
+ld=ld-linux-riscv64-lp64d.so.1
 %endif
