diff -Nur original-glibc/sysdeps/unix/sysv/linux/bits/fcntl-linux.h glibc/sysdeps/unix/sysv/linux/bits/fcntl-linux.h
--- original-glibc/sysdeps/unix/sysv/linux/bits/fcntl-linux.h   2015-08-04 23:42:21.000000000 -0700
+++ glibc/sysdeps/unix/sysv/linux/bits/fcntl-linux.h    2015-09-01 16:47:51.334378255 -0700
@@ -100,6 +100,8 @@
 # define __O_TMPFILE   (020000000 | __O_DIRECTORY)
 #endif

+/* FIXME: This needs to get upstreamed correctly. */
+#ifndef __riscv64
 #ifndef F_GETLK
 # ifndef __USE_FILE_OFFSET64
 #  define F_GETLK      5       /* Get record locking info.  */
@@ -116,6 +118,11 @@
 # define F_SETLK64     13      /* Set record locking info (non-blocking).  */
 # define F_SETLKW64    14      /* Set record locking info (blocking).  */
 #endif
+#else
+#  define F_GETLK       5       /* Get record locking info.  */
+#  define F_SETLK       6       /* Set record locking info (non-blocking).  */
+#  define F_SETLKW      7       /* Set record locking info (blocking).  */
+#endif
